### ISSUE TITLE
NPE in RTR with multiple threads for task assignment

### ISSUE THREAD

**Issue by nishantmonu51** [druid-io/druid#2842]
Multiple threads in task assignment causes NPE. 
Seems to have been introduced by https://github.com/druid-io/druid/pull/2521
```
2016-04-15T09:56:11,335 ERROR [rtr-pending-tasks-runner-2] io.druid.indexing.overlord.RemoteTaskRunner - Exception while trying to assign task: {class=io.druid.indexing.overlord.RemoteTask
Runner, exceptionType=class java.lang.NullPointerException, exceptionMessage=task, taskId=<TASKID>}
java.lang.NullPointerException: task
        at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:229) ~[druid-selfcontained-7aaf2b3.jar:7aaf2b3]
        at io.druid.indexing.overlord.RemoteTaskRunner.tryAssignTask(RemoteTaskRunner.java:638) ~[druid-selfcontained-7aaf2b3.jar:7aaf2b3]
        at io.druid.indexing.overlord.RemoteTaskRunner.access$1400(RemoteTaskRunner.java:116) ~[druid-selfcontained-7aaf2b3.jar:7aaf2b3]
        at io.druid.indexing.overlord.RemoteTaskRunner$3.call(RemoteTaskRunner.java:568) [druid-selfcontained-7aaf2b3.jar:7aaf2b3]
        at io.druid.indexing.overlord.RemoteTaskRunner$3.call(RemoteTaskRunner.java:556) [druid-selfcontained-7aaf2b3.jar:7aaf2b3]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_60]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_60]
2016-04-15T09:56:11,336 ERROR [rtr-pending-tasks-runner-2] io.druid.indexing.overlord.RemoteTaskRunner - Exception in running pending tasks: {class=io.druid.indexing.overlord.RemoteTaskRun
ner, exceptionType=class java.lang.NullPointerException, exceptionMessage=taskRunnerWorkItem}
java.lang.NullPointerException: taskRunnerWorkItem
        at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:229) ~[druid-selfcontained-7aaf2b3.jar:7aaf2b3]
        at io.druid.indexing.overlord.RemoteTaskRunner.taskComplete(RemoteTaskRunner.java:1048) ~[druid-selfcontained-7aaf2b3.jar:7aaf2b3]
        at io.druid.indexing.overlord.RemoteTaskRunner.access$1500(RemoteTaskRunner.java:116) ~[druid-selfcontained-7aaf2b3.jar:7aaf2b3]
        at io.druid.indexing.overlord.RemoteTaskRunner$3.call(RemoteTaskRunner.java:577) [druid-selfcontained-7aaf2b3.jar:7aaf2b3]
        at io.druid.indexing.overlord.RemoteTaskRunner$3.call(RemoteTaskRunner.java:556) [druid-selfcontained-7aaf2b3.jar:7aaf2b3]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_60]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_60]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_60]
```

**Comment by himanshug** [comment]
@nishantmonu51 i think i see a race between task assignment and shutdown. can you confirm if this happened in a scenario where a call was made to shutdown one or more tasks? if that is the case then having a null check in runPendingTasks() would solve the issue. also, this shouldn't have had any bad impact and overlord should have been able to make progress... did it continue to make progress after the error?
will send a PR to if what i said above is true.

**Comment by nishantmonu51** [comment]
No, I was not using shutdown. I just tried an overlord update and hit it.  
I think the race I hit was as follows - 
1) thread 1 and 2 both make a copy of pendingTasks 
2) thread 1 goes ahead and assigns the task, does putIfAbsent in tryAssignTasks, assigns the task and then finally remove the task from tryAssignTasks 
3) now thread 2 tries to assign the same task calls a tryAssignTasks.putIfAbsent which succeeds and it then tries to read the task payload which was earlier removed by thread 1.  

For the impact, i can confirm that the overlord was able to make progress and the task assignment also worked after this. I did not see any tasks failing after this. I noticed this when i saw a bunch of alert mails for this.
